master/buildbot/secrets/providers/vault_hvac.py

```diff
@@ -0,0 +1,164 @@
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+"""
+HVAC based providers
+"""
+
+
+from buildbot import config
+from buildbot.secrets.providers.base import SecretProviderBase
+
+import hvac
+
+
+class HvacAuthenticator:
+    """
+    base HVAC authenticator class
+    """
+
+    def authenticate(self, client):
+        pass
+
+
+class HvacAuthenticatorToken(HvacAuthenticator):
+    """
+    HVAC authenticator for static token
+    """
+
+    def __init__(self, token):
+        self.token = token
+
+    def authenticate(self, client):
+        client.token = self.token
+
+
+class HvacAuthenticatorApprole(HvacAuthenticator):
+    """
+    HVAC authenticator for Approle login method
+    """
+
+    def __init__(self, roleId, secretId):
+        self.roleId = roleId
+        self.secretId = secretId
+
+    def authenticate(self, client):
+        client.auth.approle.login(role_id=self.roleId, secret_id=self.secretId)
+
+
+class HvacKvSecretProvider(SecretProviderBase):
+    """
+    basic provider where each secret is stored in Vault KV secret engine
+    """
+
+    name = 'SecretInVaultKv'
+
+    def checkConfig(self, vaultServer=None, authenticator=None, secretsmount=None,
+                    apiVersion=2, pathDelimiter=':', pathEscape='\\'):
+        if not isinstance(vaultServer, str):
+            config.error("vaultServer must be a string while it is {}".format(type(vaultServer)))
+        if not isinstance(pathDelimiter, str) or len(pathDelimiter) > 1:
+            config.error("pathDelimiter must be a single character")
+        if not isinstance(pathEscape, str) or len(pathEscape) > 1:
+            config.error("pathEscape must be a single character")
+        if not isinstance(authenticator, HvacAuthenticator):
+            config.error("authenticator must be instance of HvacAuthenticator while it is {}"
+                         .format(type(authenticator)))
+        if apiVersion not in [1, 2]:
+            config.error("apiVersion {} is not supported".format(apiVersion))
+
+    def reconfigService(self, vaultServer=None, authenticator=None, secretsmount=None,
+                        apiVersion=2, pathDelimiter=':', pathEscape='\\'):
+        if secretsmount is None:
+            secretsmount = "secret"
+
+        self.secretsmount = secretsmount
+        self.pathDelimiter = pathDelimiter
+        self.pathEscape = pathEscape
+        self.authenticator = authenticator
+        self.apiVersion = apiVersion
+        if vaultServer.endswith('/'):
+            vaultServer = vaultServer[:-1]
+        self.client = hvac.Client(vaultServer)
```

Comment:
If the client class is stored in a class variable, then we could easily override it in tests and no mocking would be needed.

Reply:
I have refactored tests as sugested